### DISCUSSION TITLE
Programmatically distinguishing between production build phase and production serve phase?

### DISCUSSION THREAD

**Discussion by elmarsto** [vercel/next.js#48736] [Help]
### Summary

Hi all,

I've got a nascent blog starter brewing at https://github.com/elmarsto/substrate.git. 

It's coming along nicely -- I've successfully combined Apollo Server, Apollo Client, and appDir mode (on canary 12). 

By checking NODE_ENV I'm succesfully able to distinguish between production and development. However, I additionally need to be able to distinguish between production *at build time* (`npm run build`) and production at serve time (`npm run start`). 

How should I do this? 

I need to do this because it hits the wrong graphql endpoint during `npm run build`, which happens on my development workstation, and ought to pull from `localhost:3000/api/graphql`, and `npm run start`, which should pull from (say) `foo.bar.site/api/graphql`. 

Possible solution is to [avoid the network for local queries](https://www.apollographql.com/docs/react/performance/server-side-rendering/#avoiding-the-network-for-local-queries) by creating a separate Apollo Client -- a server-side client, if you will -- and the client-side client, used for CSR components (`<ApolloProvider ...></ApolloProvider>`, `useQuery()` and so forth).

But this seems janky. The correct solution would be to detect if the code is being run at build time or at serve time, and use a single declared instance of Apollo Client, with the URL set with the ternary operator and NODE_ENV. 

(On the *other* hand, that seems a little janky as well: flow becomes `npm run dev && npm run build` -- so the build process has a live endpoint to dial -- but I can live with that.)

**Comment by icyJoseph** [comment]
Hi,

There's an ENV var that signals the build time. 

It is called: `process.env.NEXT_PHASE` and it will be set to `phase-production-build` during build time. 

I think what you could do though, in your particular case, is to use https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration, so called runtime config.

Now with knowledge of the next/constants:

https://github.com/vercel/next.js/blob/5e6b008b561caf2710ab7be63320a3d549474a5b/packages/next/shared/lib/constants.ts#L19-L23

And combine it with the `next config` file in function mode:

```js
// next.config.js
const { PHASE_DEVELOPMENT_SERVER, PHASE_PRODUCTION_BUILD } = require('next/constants')

module.exports = (phase, { defaultConfig }) => {
  if (phase === PHASE_DEVELOPMENT_SERVER || phase === PHASE_PRODUCTION_BUILD) {
    return {
      /* development / or build time, only config options here */
    }
  }

  return {
    /* config options for all phases except development here */
  }
}
```

And you are good to go :)
